server/pulp/plugins/rsync/publish.py

```diff
@@ -0,0 +1,522 @@
+        else:
+            message = out
+        return (rv == 0, message)
+
+    def make_rsync_args(self, files_from, source_prefix, dest_prefix, exclude=None):
+        """
+        Creates a list of arguments for the rsync command
+
+        :param files_from: list of files to rsync
+        :type files_from: list of strings
+        :param source_prefix: location of files to rsync
+        :type source_prefix: str
+        :param dest_prefix: location on remote server to rsync the files to
+        :type dest_prefix: str
+        :param exclude: list of files/directories to exclude
+        :type exclude: list
+        :return: list of arguments for rsync command
+        :rtype: list
+        """
+        args = [self._CMD, self._AZQ]
+        if not self.delete:
+            args.extend(["--files-from", files_from, "--relative"])
+
+        if exclude:
+            for x in exclude:
+                args.extend(["--exclude", x])
+        args.extend(self.make_authentication())
+        if self.delete:
+            args.append("--delete")
+        if self.links:
+            args.append("--links")
+        else:
+            args.append("--copy-links")
+        args.append(source_prefix)
+        args.append(self.make_destination(dest_prefix))
+        return args
+
+    def rsync(self):
+        """
+        This method formulates the rsync command based on parameters passed in to the __init__ and
+        then executes it.
+
+        :return: (boolean indicating success or failure, stdout + stderr)
+        :rtype: tuple
+        """
+        if not self.file_list:
+            return (True, "Nothing to sync")
+        output = ""
+        tmp_path1 = os.path.join(self.get_working_dir(), str(uuid.uuid4()))
+        open(tmp_path1, 'w').write("\n".join(sorted(self.file_list)))
+        # copy files here, not symlinks
+        (is_successful, this_output) = self.remote_mkdir(self.dest_prefix)
+        if not is_successful:
+            _logger.error("Cannot create directory %s: %s" % (self.dest_prefix, this_output))
+            return (is_successful, this_output)
+        output += this_output
+        rsync_args = self.make_rsync_args(tmp_path1, str(self.src_prefix),
+                                          self.dest_prefix, self.exclude)
+        (is_successful, this_output) = self.call(rsync_args)
+        _logger.info(this_output)
+        if not is_successful:
+            _logger.error(this_output)
+            return (is_successful, this_output)
+        output += this_output
+        return (is_successful, output)
+
+    def process_main(self):
+        (successful, output) = self.rsync()
+        if not successful:
+            raise PulpCodedException(message=output)
+
+
+class Publisher(PublishStep):
+    """
+    RSync publisher class that provides the common code for publishing to remote server. Each
+    plugin needs to extend this class and add necesary logic to the _add_necesary_steps method.
```

Comment:
There are a ton of instance variables created in ``__init__`` that should be documented somewhere, such as here.